db/blip_connected_client.go

```diff
@@ -68,3 +76,68 @@ func (bh *blipHandler) handlePutRev(rq *blip.Message) error {
 	}
 	return bh.processRev(rq, &stats)
 }
+
+// Handles a Connected-Client "query" request.
+func (bh *blipHandler) handleQuery(rq *blip.Message) error {
+	if _, found := rq.Properties[QuerySource]; found {
+		return base.HTTPErrorf(http.StatusForbidden, "Only named queries allowed")
+	}
+	name, found := rq.Properties[QueryName]
+	if !found {
+		return base.HTTPErrorf(http.StatusBadRequest, "Missing 'name'")
+	}
+	query, found := bh.db.Options.ConnectedClientQueries[name]
+	if !found {
+		return base.HTTPErrorf(http.StatusNotFound, "No such query '%s'", name)
+	}
+
+	var params map[string]interface{}
+	bodyBytes, err := rq.Body()
+	if err != nil {
+		return err
+	}
+	if len(bodyBytes) > 0 {
+		if err = json.Unmarshal(bodyBytes, &params); err != nil {
+			return base.HTTPErrorf(http.StatusBadRequest, "Invalid query parameters")
+		}
+	}
+	for _, paramName := range query.Parameters {
+		if _, found := params[paramName]; !found {
+			return base.HTTPErrorf(http.StatusBadRequest, "Parameter '%s' is missing", paramName)
+		}
+	}
+
+	bh.logEndpointEntry(rq.Profile(), fmt.Sprintf("name: %s", name))
+
+	// In the N1QL string, replace "BUCKET" with the actual bucket name, wrapped in back-quotes in
+	// case it's not a valid N1QL identifier (e.g. `travel-sample`.)
+	// TODO: This is a workaround for the inability to use back-quotes for this in an SG config
+	// file, because those are already reserved for quoting long literals like the sync function.
+	statement := strings.ReplaceAll(query.Statement, "BUCKET", "`"+bh.db.Name+"`")
```

Comment:
Arbitrary keyspaces in queries feels like a footgun waiting to happen, especially given the stronger restrictions around data access with multi-tenancy and cloud.

Per-doc/channel authentication I acknowledge is still a remaining task, but this one feels like a bigger issue.

What is the approach going to be here to ensure the query being run is operating on the bucket or scope associated with the database?

Reply:
For multi-tenancy the we require db-scoped RBAC, and rely on the mobile_sync_gateway role (or similar query service roles) only being scoped to the target bucket.  That's required here, where we don't have an additional layer of protection at the SG level (since query operations are performed at the cluster level), and it's possible for an administrator to define a query that's not using `$_keyspace`.